src/features/utils/aggregate-base.js

```diff
@@ -60,17 +62,51 @@ export class AggregateBase extends FeatureBase {
         // Jserror and Metric features uses a singleton EventAggregator instead of a regular EventBuffer.
       case FEATURE_NAMES.jserrors:
       case FEATURE_NAMES.metrics:
-        this.events = this.agentRef.sharedAggregator ??= new EventStoreManager(this.agentRef, EventAggregator, entityGuid, 'shared_aggregator')
+        this.events = this.agentRef.sharedAggregator ??= new EventStoreManager(this.agentRef, EventAggregator, entityGuid, { featureName: 'shared_aggregator' })
         break
         /** All other features get EventBuffer in the ESM by default. Note: PVE is included here, but event buffer will always be empty so future harvests will still not happen by interval or EOL.
     This was necessary to prevent race cond. issues where the event buffer was checked before the feature could "block" itself.
     Its easier to just keep an empty event buffer in place. */
       default:
-        this.events = new EventStoreManager(this.agentRef, EventBuffer, entityGuid, this.featureName)
+        this.events = new EventStoreManager(this.agentRef, EventBuffer, entityGuid, this)
         break
     }
   }
 
+  handleData (data, customAttributesAreSeparate = false) {
+    /** aggregated timeslice metrics dont support byte counting at the moment */
+    if (this.events.StorageClass === EventAggregator) return this.events.add(data)
+
+    const getBytesEstimate = () => {
+      return this.events.byteSize() + (customAttributesAreSeparate ? this.agentRef.info.jsAttributesBytes : 0)
+    }
+
+    const getSMTag = (type) => {
+      return `${this.featureName}/Harvest/${type}/Seen`
+    }
+    // this will be false when the event was dropped due to exceeding max size
+    if (!this.events.add(data)) {
+      if (!this.events.isEmpty()) {
+        // if the buffer is not empty, we can try to harvest what's in there before trying to add the new event again
+        this.reportSupportabilityMetric(getSMTag('Max'), getBytesEstimate())
+        this.agentRef.runtime.harvester.triggerHarvestFor(this)
+        this.handleData(data) // try to add the event again after harvest
```

Comment:
It's not obvious the `triggerHarvest` affects or clears `this.events`, and there's potential for failure in it doing so. Meaning this could cause infinite recursion. I think this secondary add after harvest attempt should just be an add rather than a recursive call.